### DISCUSSION TITLE
Roadmap to v5 - ES6/ESM/TypeScript/ModernTheme

### DISCUSSION THREAD

**Discussion by ghiscoding** [6pac/SlickGrid#807] [Ideas]
## Project Modernization 
This is a Work in Progress and all roadblock were resolved. This next release will be filled with a **lot** of changes to modernize the project, the main changes will be the following:
1. support ES6 (and ESM), the ESM build will provide Tree Shaking support
   - this will close a very old issue #41 
2. support TypeScript and provide Type Definition (`.d.ts` files)
3. provide a more modern Styling Theme

**Note** all PRs (Pull Requests) should be done against the **[`next`](https://github.com/6pac/SlickGrid/tree/next)** branch and all Cypress E2E tests must still be passing

---

### TODOs
Below is the list of files that need to be converted to TypeScript and also other tasks that are required for our upcoming v5 release

#### Core (10/10) - all core files were migrated in PR #804)
- [x] slick.grid.js 
- [x] slick.core.js
- [x] slick.compositeeditor.js
- [x] slick.dataview.js
- [x] slick.interaction.js
- [x] slick.groupitemmetadataprovider.js
- [x] slick.editors.js
- [x] slick.formatters.js
- not sure yet what to do with the next 2 files since they uses jQuery with JSONP
  - [x] slick.remotemodel.js (ref PR #823)
  - [x] slick.remotemodel-yahoo.js (ref PR #823)

#### Controls (4/4) - all controls were migrated in PR #808)
- [x] slick.columnmenu.js
- [x] slick.columnpicker.js
- [x] slick.gridmenu.js
- [x] slick.pager.js

#### Plugins (19/19)
- [x] slick.autotooltips.js (ref PR #804)
- [x] slick.cellcopymanager.js (ref PR #804)
- [x] slick.cellexternalcopymanager.js (ref PR #808)
- [x] slick.cellmenu.js (ref PR #811)
- [x] slick.cellrangedecorator.js (ref PR #812)
- [x] slick.cellrangeselector.js (ref PR #812)
- [x] slick.cellselectionmodel.js (ref PR #812)
- [x] slick.checkboxselectcolumn.js (ref PR #813)
- [x] slick.contextmenu.js (ref PR #811)
- [x] slick.crossgridrowmovemanager.js (ref PR #817)
- [x] slick.customtooltip.js (ref PR #816)
- [x] slick.draggablegrouping.js (ref PR #814)
- [x] slick.headerbuttons.js (ref PR #811)
- [x] slick.headermenu.js (ref PR #811)
- [x] slick.resizer.js (ref PR #815)
- [x] slick.rowdetailview.js (ref PR #822)
- [x] slick.rowmovemanager.js (ref PR #817)
- [x] slick.rowselectionmodel.js (ref PR #812)
- [x] slick.state.js (ref PR #813)

#### Others
- [x] use [esbuild](https://esbuild.github.io/) to build as 3 new format types (`iife`, `esm` and `cjs`)
- [x] add [Browser-Sync](https://browsersync.io/) for development to auto-reload page after build or file changes
- [x] update all Examples to use new `dist/browser` folder
- [x] add some ESM examples with `<script type="module"> import { SlickGrid } from 'slickgrid'; </script>`
- [x] delete all images and replace them with SVGs (in pure CSS)
- [x] provide a more modern styling theme 
  - [x] provide 2 ways to customize theme styles (via SASS variables and/or CSS variables)
- [x] add TypeScript-ESLint to the build and CI workflow (ref PR #826)
- [x] get it all working in [Slickgrid-Universal](https://github.com/ghiscoding/slickgrid-universal) by using ESM imports
   - [x] all unit tests should pass
   - [x] all Cypress E2E tests should pass
- [x] remove dev deps from `package.json` that we will no longer use with our new structure (i.e. `uglifyjs`)
- [x] update the Release npm script to make it work with the new structure (ref PR #840)
   - [x] find how to update `slickgridVersion` and `v4.0.1` in comments with esbuild build (perhaps an esbuild custom plugin)
      - it turns out that reusing previous release script was enough to update version in the new location `src/slick.grid.ts` and that's it 
- [x] see if we can add VSCode debugger in the project with dev mode (ref PR #838)
- [x] upgrade to latest Cypress v12 and convert all E2E spec files to TypeScript as well (ref PR #837)
- [x] currently `slick.core.ts` and `slick.interactions.ts` were converted to TypeScript but were not converted to Classes, it might be better to do that change as well (ref PR #834)
- [x] remove all IE10-11 old code that is no longer required neither supported (i.e. `isIEPreVer9`)
- [x] ~see if we can drop `http-server` and keep only Browser-Sync for running CI Cypress E2E tests (currently only works with http-server)~
- [x] possibly add a new Example with high frequency updates to demo SlickGrid performances (i.e. Realtime Trading Platform, ref PR #835)
- [x] ~see if adding Context Sub-Menus is doable, see Discussion #806~ 
   - it's Ag-Grid code, so that won't work
- [x] we could also remove any old deprecated code like `groupBy` (which was replaced by `setGrouping`) (ref PR #833)
   https://github.com/6pac/SlickGrid/blob/d0a0e3f8ec87b4d4abbe44897fbaead55d9b5ecb/slick.dataview.js#L350-L353
- [x] throw error in `slick.remote` files if they are used without jQuery since these 2 remote API are still requiring jQuery and we won't convert them to vanilla JS.

#### For the release
- [x] write up a [Migration Guide v5](https://github.com/6pac/SlickGrid/wiki/Major-version-5.0-%E2%80%90-ES6-ESM-and-TypeScript-Support) and add link to homepage
   - see [Discussion #776](https://github.com/6pac/SlickGrid/discussions/776) for some migration detail
   - [x] mention that all source files were moved to `src/` and user **will** have to use SlickGrid files from `dist/` folder
   - [x] mention that `lib` folder got removed from npm publish
   - [x] mention all ports/wrappers using 6pac/slickgrid (i.e. Angular-Slickgrid, ...)
   - [x] provide all possible ways to import new code, via dist folder, CDN ([jsdelivr](https://www.jsdelivr.com/package/npm/slickgrid)) and npm
   - [x] add class="slickgrid-container" to migration
   - [x] removed all images, firebugx, bootstrap.css and moved select2.css to lib folder
   - [x] draggableGrouping must be binded to plugin instance (ie: `enableColumnReorder: draggableGrouping.getSetupColumnReorder.bind(draggableGrouping)`)
   - [x] add note about deprecated methods being removed (DataView `setAggregators` and `groupBy` were both replaced by `setGrouping`)
   - [x] mention new TS Generics on SlickGrid and SlickDataView
   - [x] mention about possibility to extend TS classes (because everything is `protected` which is extendable instead of `private` which is not)
   - [x] all TS Classes (core files, controls & plugins) now have the `Slick` prefix (i.e., `SlickGrid`, `SlickDataView`, `SlickColumnPicker`) when imported as CJS/ES6/ESM
      - few exceptions (without the prefix): 
         - all grouped methods: `Aggregators` `Editors`, `Formatters`
         - all enums like `ColAutosizeMode`, `RowSelectionMode`, ...

#### After the release
- [ ] maybe update [slickgrid.net](http://slickgrid.net/) website with new info and print screen of the new theme
- [ ] update [jspreadsheets](https://jspreadsheets.com/slickgrid/) details (no more jQuery) and print screens about SlickGrid
- [x] add new ESM and Icons examples to [Examples - Wiki](https://github.com/6pac/SlickGrid/wiki/Examples)
- [x] close any issues related to this new release
  - #41 -- one of SlickGrid's oldest issue requesting ES6 support
  - #255 -- add a small subset of SVGs in pure CSS and delete all images
    - why use SVG instead of Data URI? Read this article: [Probably Don’t Base64 SVG](https://css-tricks.com/probably-dont-base64-svg/)
  - #533 -- we have a new styling theme: Alpine Theme 
  - #446 -- RTL support simply by using Flex container (in the new theme)
  - #219 -- scrolling on large dataset shows blank screen
- [x] also check if PR #831 can close issue #666

**Comment by ghiscoding** [comment]
@6pac 
I'm ready for a SlickGrid 5.0 beta (or alpha), ~the only remaining and major thing left to do~ (done) is to write a migration guide (like I've done with previous versions) and I will do that in coming days. I rewrote all node scripts as `.mjs` extension so that we can use ESM (mainly the use of `import` instead of `require`) and the release script itself is pretty much the same except that the build is now from esbuild (from `build.mjs`) instead of a minify script (which is now gone). I have updated some npm packages again last night, like TypeScript and Cypress, so make sure that you run npm install before you do anything. You could also run a dry-run (preview) of the release by running the preview script (if you do, just make sure to undo any file changes from git before you later run the official release afterward). I also added a minimum requirement of Node 16 to run the project in the `package.json` file, even though Node 16 will become EOL next Monday Sept. 11 so you might as well install Node 18. 

#### Summary of steps for a 5.0 Beta release:
1. Make sure to have Node 16 or higher (probably better with Node 18 since 16 is now EOL)
2. Fetch latest code from the `next` branch
3. Install latest dependencies
```sh
npm install
```
4. Optionally execute a dry-run (preview) release (it execute a full build and then preview a release)
   - if you do a dry-run, then make sure to rollback git changes once you are satisfied with the preview
```sh
npm run preview:release
```
5. Run a full release of 5.0 Beta (or alpha release)
```sh
npm run release # choose bump option 6 or higher
```

Cheers

EDIT

I wrote the new [Migration 5.0](https://github.com/6pac/SlickGrid/wiki/Major-version-5.0-%E2%80%90-ES6-ESM-and-TypeScript-Support)

**Comment by 6pac** [comment]
Hey Ghislain, sorry about dragging my heels, I’ve been sick :-(
TBH I’ve been wondering if you should just take over the repo at this point – I barely understand all the packaging requirements, and if I test them it’s going to be extremely cosmetic.
Would you be happy with that, and if so, what would we need to do to make it happen?
B
From: Ghislain B. ***@***.***>
Date: Saturday, 9 September 2023 at 12:14 am
To: 6pac/SlickGrid ***@***.***>
Cc: Ben McIntyre ***@***.***>, Mention ***@***.***>
Subject: Re: [6pac/SlickGrid] Roadmap to v5 - ES6/ESM/TypeScript/ModernTheme (WIP 👷🏻) (Discussion #807)

@6pac<https://github.com/6pac>
I'm ready for a SlickGrid 5.0 beta (or alpha), the only remaining and major thing left to do is to write a migration guide (like I've done with previous versions) and I will do that in coming days. I rewrote all node scripts as .mjs extension so that we can use ESM (mainly the use of import instead of require) and the release script itself is pretty much the same except that the build is now from esbuild (from build.mjs) instead of a minify script (which is now gone). I have updated some npm packages again last night, like TypeScript and Cypress, so make sure that you run npm install before you do anything. You could also run a dry-run (preview) of the release by running the preview script (if you do, just make sure to undo any file changes from git before you later run the official release afterward). I also added a minimum requirement of Node 16 to run the project in the package.json file, even though Node 16 will become EOL next Monday Sept. 11 so you might as well install Node 18.

Summary of steps for a 5.0 Beta release:

  1.  Make sure to have Node 16 or higher
  2.  Fetch latest code from the next branch
  3.  Install latest dependencies

npm install

  1.  Optionally execute a dry-run (preview) release (it execute a full build and then preview a release)

     *   if you do a dry-run, then make sure to rollback git changes once you are satisfied with the preview

npm run preview:release

  1.  Run a full release of 5.0 Beta (or alpha release)

npm run release # choose bump option 6 or higher

Cheers

**Comment by ghiscoding** [comment]
@6pac oh sorry to hear that, I hope you recover quickly and enjoy the upcoming Spring on your side :)
As for the release, I wouldn't mind to take it over but at this point I cannot since I do not have access to the NPM, you would have to add me to the npm project (below is who currently has access to publish on npm, as you can see I'm not in this list)

![image](https://github.com/6pac/SlickGrid/assets/643976/7e8f6d59-9fda-47b5-8aa7-9ac1e3381b3e)

**Comment by 6pac** [comment]
Thanks @ghiscoding, I can add your NPM user to the approved list, do you have a username for NPM, or if not, could you register?

**Reply by ghiscoding** [reply]
@6pac yes it's the same as GitHub, `ghiscoding`

**Reply by 6pac** [reply]
should have thought of that! you've been invited as a maintainer

**Reply by ghiscoding** [reply]
got the invite thanks... wish you well soon, I didn't catch covid yet (at least I think) but I'm sure it's just a matter of time

**Reply by 6pac** [reply]
great. you're onto it, I see. that will be a lot easier for both of us ;-)

**Reply by ghiscoding** [reply]
official v5 is out, I made a couple of by releasing alpha versions without the alpha tag (oops) so I decided to go ahead with the official v5 to erase the mistake haha

There's maybe 1 task left that has to be done by you, it should be easy but is not urgent

> * maybe update [slickgrid.net](http://slickgrid.net/) website with new info and print screen of the new theme

**Reply by ghiscoding** [reply]
@6pac 
I'm not sure if you saw my last comment, in case you didn't, there's this task left for you to do eventually (not urgent though) 😉 

> * maybe update [slickgrid.net](http://slickgrid.net/) website with new info and print screen of the new theme

oh can you also update the GitHub project link, I think the link should be either pointing to the **Examples** page OR your slickgrid.net website

![image](https://github.com/6pac/SlickGrid/assets/643976/7a5c56b9-f5d5-4ec2-9317-6a88bfe2489f)
